src/library/scala/collection/IndexedSeq.scala

```diff
@@ -25,6 +25,28 @@ trait IndexedSeq[+A] extends Seq[A]
   override protected[this] def stringPrefix: String = "IndexedSeq"
 
   override def iterableFactory: SeqFactory[IndexedSeq] = IndexedSeq
+
+  override def foldLeft[B](z: B)(f: (B, A) => B): B = {
+    var b = z
+    var i = 0
+    while (i < length) {
+      val a = apply(i)
+      i += 1
+      b = f(b, a)
+    }
+    b
+  }
+
+  override def foldRight[B](z: B)(f: (A, B) => B): B = {
+    var b = z
+    var i = length
+    while (i > 0) {
+      i -= 1
+      val a = apply(i)
```

Comment:
This surely has much better memory usage than reversing and folding left, but again, indexing may be worse than a `reverseIterator`, so that should be used instead.

Note that the iterator should be used raw:
```scala
val r = reverseIterator()
var b = z
while (r.hasNext) b = f(b, r.next)
b
```